File: java/src/binary_heap/HeapSetTest.java
package binary_heap;

import org.junit.Assert;
import org.junit.Test;

import java.util.Arrays;

public class HeapSetTest {
	@Test
	public void testRandomCases() {
		for (int test = 0; test < 1_000; test++) {
			var td = new randomTestData();

			Assert.assertTrue(randomTestData.minHeapIsOK(td.heap));

			var actual = td.heap.clone();
			HeapSet.heapSet(actual, td.randomIndex, td.randomValue);

			td.heap[td.randomIndex] = td.randomValue;

			if (!randomTestData.minHeapIsOK(actual)) {
				Assert.fail(
					String.format("heapSet(%s, %s, %s): not a valid heap after the call:\n%s",
						Arrays.toString(td.heap), td.randomIndex, td.randomValue, Arrays.toString(actual)
					)
				);
			}

			if (!randomTestData.haveSameElements(actual, td.heap)) {
				Assert.fail(
					String.format("heapSet(%s, %s, %s): not same elements after the call:\n%s",
						Arrays.toString(td.heap), td.randomIndex, td.randomValue, Arrays.toString(actual)
					)
				);
			}
		}
	}
}

File: go/binary_heap/heap_set_test.go
package binary_heap

import (
	"github.com/stretchr/testify/assert"
	"testing"
)

func TestHeapSet_RandomCases(t *testing.T) {
	for range 1_000 {
		td := newRandomTestData()

		assert.True(t, minHeapIsOK(td.heap))

		actual := append([]int{}, td.heap...)
		heapSet(actual, td.randomIndex, td.randomValue)

		td.heap[td.randomIndex] = td.randomValue

		if !minHeapIsOK(actual) {
			assert.FailNowf(t, "",
				"heapSet(%v, %v, %v): not a valid heap after the call:\n%v",
				td.heap, td.randomIndex, td.randomValue, actual,
			)
		}

		if !haveSameElements(actual, td.heap) {
			assert.FailNowf(t, "",
				"heapSet(%v, %v, %v): not same elements after the call:\n%v",
				td.heap, td.randomIndex, td.randomValue, actual,
			)
		}
	}
}
